examples/MvcExample/Controllers/HomeController.cs

```diff
@@ -84,34 +85,49 @@ public async Task<IActionResult> Complete([FromQuery(Name = "payment_id")]string
 
             var apiResponse = await _truelayer.Payments.GetPayment(paymentId);
 
-            IActionResult Failed(string status)
+            IActionResult Failed(string status, OneOf<Provider.UserSelected, Provider.Preselected>? providerSelection)
             {
                 ViewData["Status"] = status;
+
+                SetProviderAndSchemeId(providerSelection);
                 return View("Failed");
             }
 
-            IActionResult Success(PaymentDetails payment)
+            IActionResult Success(PaymentDetails payment, OneOf<Provider.UserSelected, Provider.Preselected> providerSelection)
             {
                 ViewData["Status"] = payment.Status;
+                SetProviderAndSchemeId(providerSelection);
                 return View("Success");
             }
 
-            IActionResult Pending(PaymentDetails payment)
+            IActionResult Pending(PaymentDetails payment, OneOf<Provider.UserSelected, Provider.Preselected> providerSelection)
             {
                 ViewData["Status"] = payment.Status;
+                SetProviderAndSchemeId(providerSelection);
                 return View("Success");
             }
 
+            void SetProviderAndSchemeId(OneOf<Provider.UserSelected, Provider.Preselected>? providerSelection)
+            {
+                (string providerId, string schemeId) = providerSelection?.Match(
+                    userSelected => (userSelected.ProviderId, userSelected.SchemeId),
+                    preselected => (preselected.ProviderId, preselected.SchemeId)
+                ) ?? ("unavailable", "unavailable");
+
+                ViewData["ProviderId"] = providerId;
+                ViewData["SchemeId"] = schemeId;
+            }
+
             if (!apiResponse.IsSuccessful)
-                return Failed(apiResponse.StatusCode.ToString());
+                return Failed(apiResponse.StatusCode.ToString(), null!);
 
             return apiResponse.Data.Match(
-                authRequired => Failed(authRequired.Status),
-                authorizing => Pending(authorizing),
-                authorized => Success(authorized),
-                success => Success(success),
-                settled => Success(settled),
-                failed => Failed(failed.Status)
+                authRequired => Failed(authRequired.Status, apiResponse.Data.AsT0.PaymentMethod.AsT0.ProviderSelection),
```

Comment:
There's no need to cast `Data`, the Match statement is intended to handle that, so you could do `authRequired.PaymentMethod....` and so on in the following branches.
Generally I would avoid the use of `AsTx` in application code unless it's absolutely necessary or there's a guard in place to ensure that it is of that type. Otherwise you can run the risk of an InvalidOperationException. It's also a bit opaque as to what T0 actually is.

Reply:
Good point on the cast for `Data`, thanks. 
As for `AsTx` I generally do the same but this is just an example project, not a production one so I thought it would be ok anyway.

Reply:
Addressed in [86d5849](https://github.com/TrueLayer/truelayer-dotnet/pull/136/commits/86d5849a14c112ce1520ee1476c6d5d51f90297d)